File: python/m3/m3.3_homework_filled.py
# python/m3/m3.3_homework_filled.py
"""Reference copy of m3.3_homework.py with TODOs 1 and 2 filled in so you
can run it end to end and see what "done" looks like. This is just one
possible answer, so yours might be different. Explore!"""

from deepagents import create_deep_agent
from deepagents.backends import CompositeBackend, StateBackend, StoreBackend
from deepagents.backends.utils import create_file_data
from langgraph.store.memory import InMemoryStore

from models import model

store = InMemoryStore()
memory_path = "/memories/AGENTS.md"
store_memory_path = "/AGENTS.md"

CONTEXT_A = {"workspace_id": "homework", "user_id": "u_you"}
CONTEXT_B = {"workspace_id": "homework", "user_id": "u_teammate"}


def namespace_from_context(context):
    return ("memory", context["workspace_id"], context["user_id"])


def memory_namespace(runtime):
    return namespace_from_context(runtime.context)


# TODO 1 filled in
def build_seed_memory_a() -> str:
    return """\
# Houseplant Notes

## Watering
- The fiddle-leaf fig gets watered every 10 days, not on a fixed weekday;
  check the top inch of soil first.
- The succulents on the windowsill only get watered when the soil is
  completely dry, roughly every 2-3 weeks.

## Light
- The pothos and snake plant tolerate low light and live in the hallway.
- Everything else needs the south-facing window.
"""


def build_seed_memory_b() -> str:
    return """\
# Herb Garden Notes

## Watering
- Basil and mint want consistently moist soil; check daily in summer.
- Rosemary is drought-tolerant; only water when the top two inches are dry.

## Light
- All three herbs live on the kitchen windowsill, which gets morning sun.
"""


store.put(namespace_from_context(CONTEXT_A), store_memory_path, create_file_data(build_seed_memory_a()))
store.put(namespace_from_context(CONTEXT_B), store_memory_path, create_file_data(build_seed_memory_b()))

agent = create_deep_agent(
    model=model,
    name="Homework_Memory_Agent",
    backend=CompositeBackend(
        default=StateBackend(),
        routes={"/memories/": StoreBackend(namespace=memory_namespace)},
    ),
    store=store,
    memory=[memory_path],
    system_prompt="You are a helpful personal assistant for this project.",
)


# TODO 2 filled in
RECALL_QUESTION = "How often should I water the fiddle-leaf fig, and where does the pothos live?"
REMEMBER_MESSAGE = (
    "Remember: I just repotted the fiddle-leaf fig, so skip watering it for "
    "the next 3 weeks while the roots settle. Update your memory."
)
LEAK_CHECK_QUESTION = "How often should I water the fiddle-leaf fig, and where does the pothos live?"

# 1. Context A recalls from its own seed.
result_a1 = agent.invoke({"messages": [{"role": "user", "content": RECALL_QUESTION}]}, context=CONTEXT_A)
print("--- Context A, Question 1 ---")
print(result_a1["messages"][-1].content)

# 2. Context A learns a new, distinctive fact.
result_a2 = agent.invoke({"messages": [{"role": "user", "content": REMEMBER_MESSAGE}]}, context=CONTEXT_A)
print("\n--- Context A, Question 2 (remember) ---")
print(result_a2["messages"][-1].content)

# 3. Context B asks the same question. It should NOT see anything from A.
result_b = agent.invoke({"messages": [{"role": "user", "content": LEAK_CHECK_QUESTION}]}, context=CONTEXT_B)
print("\n--- Context B, leak-check question ---")
print(result_b["messages"][-1].content)

memory_a = store.get(namespace_from_context(CONTEXT_A), store_memory_path).value["content"]
memory_b = store.get(namespace_from_context(CONTEXT_B), store_memory_path).value["content"]
print("\n--- Context A's stored AGENTS.md ---")
print(memory_a)
print("\n--- Context B's stored AGENTS.md ---")
print(memory_b)

if memory_a == memory_b:
    print("\nISOLATION FAILED: both contexts share identical stored memory.")
else:
    print("\nStored memories differ between contexts, as expected.")

File: typescript/m3/m3.3_homework_filled.ts
// typescript/m3/m3.3_homework_filled.ts
/**
 * Reference copy of m3.3_homework.ts with TODOs 1 and 2 filled in so you
 * can run it end to end and see what "done" looks like. This is just one
 * possible answer, so yours might be different. Explore!
 */

import {
  createDeepAgent,
  CompositeBackend,
  StateBackend,
  StoreBackend,
  type FileData,
  type StoreBackendContext,
} from "deepagents";
import { InMemoryStore } from "@langchain/langgraph";

import { model } from "../models.js";

// deepagents does not export create_file_data, so build the FileDataV2 shape by hand.
function createFileData(content: string): FileData {
  const now = new Date().toISOString();
  return { content, mimeType: "text/markdown", created_at: now, modified_at: now };
}

const store = new InMemoryStore();
const memoryPath = "/memories/AGENTS.md";
const storeMemoryPath = "/AGENTS.md";

const CONTEXT_A = { workspace_id: "homework", user_id: "u_you" };
const CONTEXT_B = { workspace_id: "homework", user_id: "u_teammate" };

function namespaceFromContext(context: { workspace_id: string; user_id: string }): string[] {
  return ["memory", context.workspace_id, context.user_id];
}

function memoryNamespace(context: StoreBackendContext): string[] {
  const configurable = (context.config?.configurable ?? {}) as {
    workspace_id: string;
    user_id: string;
  };
  return namespaceFromContext(configurable);
}

// TODO 1 filled in
function buildSeedMemoryA(): string {
  return `# Houseplant Notes

## Watering
- The fiddle-leaf fig gets watered every 10 days, not on a fixed weekday;
  check the top inch of soil first.
- The succulents on the windowsill only get watered when the soil is
  completely dry, roughly every 2-3 weeks.

## Light
- The pothos and snake plant tolerate low light and live in the hallway.
- Everything else needs the south-facing window.
`;
}

function buildSeedMemoryB(): string {
  return `# Herb Garden Notes

## Watering
- Basil and mint want consistently moist soil; check daily in summer.
- Rosemary is drought-tolerant; only water when the top two inches are dry.

## Light
- All three herbs live on the kitchen windowsill, which gets morning sun.
`;
}

await store.put(
  namespaceFromContext(CONTEXT_A),
  storeMemoryPath,
  createFileData(buildSeedMemoryA())
);
await store.put(
  namespaceFromContext(CONTEXT_B),
  storeMemoryPath,
  createFileData(buildSeedMemoryB())
);

const agent = createDeepAgent({
  model,
  name: "Homework_Memory_Agent",
  backend: new CompositeBackend(new StateBackend(), {
    "/memories/": new StoreBackend({ namespace: memoryNamespace }),
  }),
  store,
  memory: [memoryPath],
  systemPrompt: "You are a helpful personal assistant for this project.",
});

// TODO 2 filled in
const RECALL_QUESTION =
  "How often should I water the fiddle-leaf fig, and where does the pothos live?";
const REMEMBER_MESSAGE =
  "Remember: I just repotted the fiddle-leaf fig, so skip watering it for " +
  "the next 3 weeks while the roots settle. Update your memory.";
const LEAK_CHECK_QUESTION =
  "How often should I water the fiddle-leaf fig, and where does the pothos live?";

const configA = { configurable: CONTEXT_A };
const configB = { configurable: CONTEXT_B };

// 1. Context A recalls from its own seed.
const resultA1 = await agent.invoke(
  { messages: [{ role: "user", content: RECALL_QUESTION }] },
  configA
);
console.log("--- Context A, Question 1 ---");
console.log(resultA1.messages.at(-1)?.content);

// 2. Context A learns a new, distinctive fact.
const resultA2 = await agent.invoke(
  { messages: [{ role: "user", content: REMEMBER_MESSAGE }] },
  configA
);
console.log("\n--- Context A, Question 2 (remember) ---");
console.log(resultA2.messages.at(-1)?.content);

// 3. Context B asks the same question. It should NOT see anything from A.
const resultB = await agent.invoke(
  { messages: [{ role: "user", content: LEAK_CHECK_QUESTION }] },
  configB
);
console.log("\n--- Context B, leak-check question ---");
console.log(resultB.messages.at(-1)?.content);

const memoryA = ((await store.get(namespaceFromContext(CONTEXT_A), storeMemoryPath))!
  .value as FileData).content;
const memoryB = ((await store.get(namespaceFromContext(CONTEXT_B), storeMemoryPath))!
  .value as FileData).content;
console.log("\n--- Context A's stored AGENTS.md ---");
console.log(memoryA);
console.log("\n--- Context B's stored AGENTS.md ---");
console.log(memoryB);

if (memoryA === memoryB) {
  console.log("\nISOLATION FAILED: both contexts share identical stored memory.");
} else {
  console.log("\nStored memories differ between contexts, as expected.");
}
